### ISSUE TITLE
Handle up to 255 full Unicode Codepoints (not 255 bytes)

### ISSUE THREAD

**Issue by daniel-j-h** [Project-OSRM/osrm-backend#2844]
Issue for test case added in https://github.com/Project-OSRM/osrm-backend/pull/2821.

The OSM spec says "up to 255 full unicode chars", so we should be able
to handle 255 x 🐼 --- but we don't.

The internal limit is on 255 _bytes_ in the `std::string`, but there can
be e.g. up to four bytes in a utf8 encoding.

References:
- http://wiki.openstreetmap.org/wiki/API_v0.6#Tags
- http://www.fileformat.info/info/unicode/char/1f43c/browsertest.htm
- https://en.wikipedia.org/wiki/UTF-8#Description
- https://github.com/Project-OSRM/osrm-backend/blob/807aa71a7c80b714b7a89db9c99c15720572ba25/include/util/range_table.hpp#L28

**Comment by MoKob** [comment]
The feature to check for this is in (https://github.com/Project-OSRM/osrm-backend/pull/2821), now we just need to handle it.

**Comment by daniel-j-h** [comment]
I looked into the range table being the limiting factor for how we ultimately store strings. @MoKob pointed out the prefix sum being the main issue here. Playing with it for a bit and switching from a `char` to a larger type now lets the intersection classification I/O code fail.

```
terminate called after throwing an instance of 'osrm::util::exception'
  what():  Failed to read bearing values from /tmp/osrm-backend/test/cache/ec565977fc21e9bcff967d61a3bd71e09a5b7460_c499d2c4587859ff929c81688cdef37d5dbc70bd_c12f16d2fcd912fb00db7ff4f55d4106f9d0cc57.osrm.icd
```

**Comment by MoKob** [comment]
@daniel-j-h this is also reading in a range table. So if you break the range table, you have to adjust the code for reading the intersection_classification in as well.

**Comment by TheMarex** [comment]
The reason why the range table operates on bytes is that this enables vectorized computation of the prefix-sum. You would need to check not only the memory impact (for using 16bit) but also the performance impact on the range table.

The specification might say 255 Unicode chars, but what do we actually hit on a planet extract?

As a side note there is a wealth of research for efficiently encoding delta-int values like these. Most of them are based on varint encodings - we might give this a spin and drop the hard coded length limit altogether. For example Folly has an implementation for GroupVarint https://github.com/facebook/folly/blob/master/folly/docs/GroupVarint.md but there are numerous other open source libs for this.

**Comment by oxidase** [comment]
@TheMarex i have made some modified version of varint-GB with 4-byte descriptor https://github.com/oxidase/home/blob/master/notes/topics/c%2B%2B/indexed_data.cpp 
Advantages over fixed block size:
* lengths range 0-16777215 
* better compression of data with over 25% zeros, because 0 is encoded with 2 bits  instead of 8
As a negative point: no more easy compiler-based SSE parallelization and some overhead at the descriptor decoding. Probably it can be optimized via pshufb instruction, but it will require at least 256*16 shuffle table and i think 4 memory accesses to the table to decode one descriptor will diminish  SSE speed-up.

If idea is ok, i can adjust it to the osrm codebase.

**Comment by TheMarex** [comment]
@oxidase looks great! Replacing the RangeTable with that should do nicely. There are probably some other places that would benefit from a varint encoding, so this will be a worthwhile addition to the codebase.

My only concern is speed, a micro-benchmark should be enough just so that we get an idea of the slow-down. Access to name/ref/pronunciation values is pretty rare (roughly in the number of instructions, so I would estimate below 500 per query). This should get you an idea of the slow down we would see for real (very long) requests.

**Comment by TheMarex** [comment]
This has been fixed by #3585
